ccl/sqlproxyccl: complete connection migration support in the forwarder

jaylim-crl opened the pull request:
#### ccl/sqlproxyccl: add last message details to the forwarder's processor

This commit adds last message details to the forwarder's processor, and in
particular, lastMessageType and lastMessageTransferredAt. The latter is
implemented using a simple logical clock, which will be used to determine
ordering of events. These two will be used during connection migration to
determine a safe transfer point.

At the same time, we plumb the connector and metrics into the processor, which
will be used in subsequent commits.

Release justification: Low risk, sqlproxy-only change.

Release note: None

#### ccl/sqlproxyccl: support waitResumed on the processor to block until resumption

Previously, there could be a race where suspend() was called right after
resuming the processors. If the processor goroutines have not started, suspend
will implicitly return, leading to a violation of an invariant, where we want
the processors to be suspended before proceeding. This commit adds a new
waitResumed method on the processor that allows callers to block until the
processors have been resumed.

Release justification: sqlproxy-only change.

Release note: None

#### ccl/sqlproxyccl: complete connection migration support in the forwarder

Informs cockroachdb#76000.

This commit completes the connection migration feature in the the forwarder
within sqlproxy. The idea is as described in the RFC.

A couple of new sqlproxy metrics have been added as well:
- proxy.conn_migration.success
- proxy.conn_migration.error_fatal
- proxy.conn_migration.error_recoverable
- proxy.conn_migration.attempted

For more details, see metrics.go in the sqlproxyccl package.

Release justification: This completes the first half of the connection
migration feature. This is low risk as part of the code is guarded behind the
connection migration feature, which is currently not being used in production.
To add on, CockroachCloud is the only user of sqlproxy.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76805)
<!-- Reviewable:end -->

jaylim-crl reviewed (commented):
The structure for the connection migration work is there. Note that this is untested, and was inspired mostly from the initial prototype. Putting this up in case anyone has thoughts on the direction. I'll be adding tests next.

When it comes to the forwarder's API, I plan to have a Forwarder interface that implements just two methods: `RequestTransfer` and `Close`. With the current design, callers wouldn't know whether the transfer has failed or succeeded after calling `RequestTransfer`. I think that is fine for now, and if we really need to know the last transfer error, we could address that in the upcoming RFC. That said, the initial thought is to have the forwarder _notify_ a "forwarder tracking" component (i.e. map of pods to forwarders) if it connects/disconnects from `serverConn`. The proxy can then check which pod the forwarder is assigned to (versus the previous pod) to know whether the transfer has succeeded. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MwFOvRJ1VOOr16lmRhb:bxev2y1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 52 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFO_ka7WzDvvDcSIUP:-MwFO_kb310jFSYLbkZ2:b2wk8d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4797bdfde552b2db38ee66c0f60c1a0c41dbb9c1/pkg/ccl/sqlproxyccl/forwarder.go#L52)):*
> ```Go
> )
> 
> var (
> ```

These error messages might be unnecessary, at least for the ones where they are only used once. We could just do return `errors.New` there if we wanted to.


<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/ccl/sqlproxyccl/forwarder.go:
```diff
@@ -130,37 +260,548 @@ func (f *forwarder) Close() {
+			}
+
+		case stateTransferInProgress:
+			if err := f.runTransferWithTimeout(
```
idea: Instead of implementing the forwarding goroutines as a state machine, initiating the transfer could ask both of the goroutines to terminate. That allows us to implement transfer the same way authenticate is implemented. As a single function and goroutine. Once the transfer is canceled or completed, the goroutine managing the transfer could restart the forwarding goroutines.

IMO the main advantage to implementing it by stopping the forwarding goroutines is we can keep the steady state code simple and we can group all of the transfer logic together.

jaylim-crl reviewed (commented):
TFTRs!

This draft PR is ready for another look. I added more tests, and changed how we handle timeout handlers. I also tested this, and it works as expected.

This test is still WIP, and is not included in this PR. Will update that later today. Here, we call `RequestTransfer` in a loop, and constantly query the SQL pod details. We can see that we alternate between pods, which implies that connection migration works:

```
=== RUN   TestConnectionMigration/successful_connection
{postgresql://root@127.0.0.1:39917/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:36529/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:39917/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:36529/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:39917/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:36529/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:39917/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:36529/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:39917/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
{postgresql://root@127.0.0.1:36529/defaultdb?sslcert=test_certs%2Fclient.root.crt&sslkey=test_certs%2Fclient.root.key&sslmode=verify-full&sslrootcert=test_certs%2Fca.crt true}
```

There are three more items to address:
1. Implement deserialize_session. Given that this PR is huge, I might just push that to a subsequent PR.
2. Perhaps move the changes in authentication.go in another PR. One thing that was overlooked is that the server still sends back initial connection messages in the token-based auth approach, so we'd have to slurp those messages.
3. Add `TestConnectionMigration` (which is an end-to-end test) to this PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-Mwhb94i9bCtE3PIPWVF:b-qb1r74)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r2-269:-Mwhbyby8O4t87vp7GWd:b-wc8gdm) ([raw file](https://github.com/cockroachdb/cockroach/blob/009f2d62292436724ab005606f3a6efb5d5fd4db/pkg/ccl/sqlproxyccl/forwarder.go#L269)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Can we do this without atomics?  I do not like using atomics for control flow. Getting atomic programming right is infamously difficult. In this case there is a race condition that occurs if RequestTransfer is called concurrently and it's not clear to me how to fix it. At a minimum CompareAndSwap needs to be used as the guard for this state transfer. But then there is a race with setting transferKey and transferCh.
</blockquote></details>

Done - good catch re: concurrent calls. As discussed, I have moved this to use locks. My benchmarks show that using locks within a 30s workload run only slows the overall time by a few hundred ms, so I'd be fine with using locks everywhere instead of a mixture of both.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 440 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r2-440:-MwhcHSe4EcTz6gUD7S7:befkccy) ([raw file](https://github.com/cockroachdb/cockroach/blob/009f2d62292436724ab005606f3a6efb5d5fd4db/pkg/ccl/sqlproxyccl/forwarder.go#L440)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

idea: Instead of implementing the forwarding goroutines as a state machine, initiating the transfer could ask both of the goroutines to terminate. That allows us to implement transfer the same way authenticate is implemented. As a single function and goroutine. Once the transfer is canceled or completed, the goroutine managing the transfer could restart the forwarding goroutines.

IMO the main advantage to implementing it by stopping the forwarding goroutines is we can keep the steady state code simple and we can group all of the transfer logic together.
</blockquote></details>

As discussed, this would increase the overall transfer latency since we'd have to wait for the server-to-client processor to stop before sending the transfer state message, so I kept the current approach. I like the idea as it makes things cleaner though, but the tradeoff wouldn't be worth it.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 514 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r2-514:-MwhcUvy38XIfD3o5QaY:bvrzgb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/009f2d62292436724ab005606f3a6efb5d5fd4db/pkg/ccl/sqlproxyccl/forwarder.go#L514)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

This logic can be simplified by context.WithTimeout.

transferCtx, cancel := context.WithTimeout(timeout, ctx)
defer cancel()
err := transfer(trasferCtx)
if transferCtx.Err() && !ctx.Err() {
  return errTransferTimeout
}
return err

Alternatively, if we do not specifically care about errTransferTimeout, it can use context.WithTimeout.
</blockquote></details>

Hm, this probably isn't relevant anymore since I moved it elsewhere.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
> There are three more items to address:
> 
> 1 .Implement deserialize_session. Given that this PR is huge, I might just push that to a subsequent PR.
> 2. Perhaps move the changes in authentication.go in another PR. One thing that was overlooked is that the server still sends back initial connection messages in the token-based auth approach, so we'd have to slurp those messages.
> 3. Add TestConnectionMigration (which is an end-to-end test) to this PR.

(2) and (3) are done. I've decided to just leave them in this PR, but happy to break them down into separate PRs if reviewers prefer otherwise. (1) will definitely come in a later PR.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MwjNFJAA27XNnA5g7ZP:ba0ntzo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 629 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwjNcdREnHFZ1IBwIz6:-MwjNcdREnHFZ1IBwIz7:b-5d9dx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/87b8ea54173ef954e4c0b08e6279e232c0113530/pkg/ccl/sqlproxyccl/forwarder.go#L629)):*
> ```Go
> // NOTE: This should only be called during a transfer process. We assume that
> // transferCloserCh has already been initialized.
> func (f *forwarder) runTransferTimeoutHandler(waitTimeout time.Duration) {
> ```

The end-to-end transfer process works, but I'm not happy with this function. The unfortunate thing is that we unblock clients through context, but for servers, we'd do that by closing the forwarder. This goroutine is needed to ensure that we can terminate if we timeout when sending the `SHOW TRANSFER STATE` query to the server. Also perhaps I ran into a rabbit hole with all the `transferConnRecoverable` stuff, but I don't see a better way to address that. Happy to discuss.

`closer` is needed instead of relying on `transferCtx.Done()` to ensure that we can distinguish between a transfer timeout, and an actual transfer completion. If we block on `transferCtx.Done()` in the client-to-server processor, then when the timeout handler gets triggered, we will unblock automatically, which is incorrect.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
FYI, I extracted some of the work here in https://github.com/cockroachdb/cockroach/pull/77109. I'll do the same for the auth stuff.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-Mwx8WzrDzR4j1k8OhzC:b-1hvrvg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)



<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/ccl/sqlproxyccl/forwarder.go:
```diff
@@ -130,37 +276,428 @@ func (f *forwarder) Close() {
+
+	// Connect to a new SQL pod.
+	//
+	// TODO(jaylim-crl): There is a possibility where the same pod will get
```
Consider adding the target pod as an argument to RequestTransfer. That way a central component gets to choose where the connections go.

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-Mx1W_QM1tr1HeDYrIcA:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-151:-Mx1Ulo355uui1kiwloP:b1hjcgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L151)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

transferCtx does not need to be on the mutex guarded struct. It can be initialized in processTransfer and passed as arguments to the functions called by processTransfer. 
</blockquote></details>

It's currently set within `runTransferTimeoutHandler`. If we initialized in `processTransfer`, we're changing our definition of the timeout here - maybe that's OK. Previously, the transfer timeout starts right when we send the SHOW TRANSFER STATE query, which aligns well with the `stateTransferInProgress` state. With this change, it starts from when we called RequestTransfer successfully.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 156 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-156:-Mx1RMyV5SE8UzRvbG9E:bab7pqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L156)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Currently this is set only in the stateTransferRequested phase after the session serialization completes. What do you think of removing the bool flag and folding it into the states? I'm thinking of something like:

```
	stateReady int = iota
	stateTransferRequested
    stateSerializingSession
	stateStartingConnection
```
where stateTransferInProgress is split into stateSerializingSession and stateStartingConnection. The state transition happens where the code sets transferConnRecoverable = true.

</blockquote></details>

That should work. How does the following sound: `stateTransferSessionSerialization`, and `stateTransferSessionDeserialization`? The latter includes trying to connect to a new SQL pod as well.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 600 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-600:-Mx1SC0Z4mbUcUZCKvyB:b-k2bnml) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L600)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Consider adding the target pod as an argument to RequestTransfer. That way a central component gets to choose where the connections go.
</blockquote></details>

We could definitely do that to make the API more robust, but I don't see a strong use case to that yet. The current intention is to reuse the existing selection algorithm for the connection migration work: https://github.com/cockroachdb/cockroach/blob/94c73132dba52afd20544c2792fdbdcd9ccddb77/pkg/ccl/sqlproxyccl/tenant/pod.go#L11-L23

That said, adding a _target_ pod makes it more complex compared to a _preferred_ pod since we now need to handle the case where the pod couldn't take in more connections. 
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 139 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwyybLiqoDcTTUX-fy3-r8-139:-Mx1TT0V9NZLEqKMgC8Y:b-1o5owm) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/metrics.go#L139)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

What do you think of renaming proxy.conn_migration.error -> proxy.conn_migration.error_fatal and proxy.conn_migration.fail -> proxy.conn_migration.error_recoverable?
</blockquote></details>

I can do that. No strong preferences here.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwyybLiqoDcTTUX-fy3-r8-151:-Mx1U3WH-XbvkiI8-WzO:b-c9y872) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/metrics.go#L151)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

I think we should fold the timeout_closed and timeout_recoverable metrics into error_count and failed_count. We can use logging to indicate the specific cause of the failure.
</blockquote></details>

Hm, let me look into this again. We may have to change where we increment the metric for this to work.
___
*[pkg/ccl/sqlproxyccl/proxy_handler_test.go, line 795 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwjMLuz2qUeQDPEZhi0-r8-795:-Mx1UTSSA86tWpVcav5Z:b-c4mx75) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/proxy_handler_test.go#L795)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

Why do we need this afterForward testing hook? Can we delete it?
</blockquote></details>

We need this to retrieve the forwarder instance, and since we don't have a balancer component (or something that holds a list of pod to forwarders), we'd have to rely on this hook. We should be able to remove this hook once we have that said component. Happy to discuss if you have other suggestions.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
All comments have been addressed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-Mx2jzi6BrpCD-Psu6AF:bhvpn6n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-151:-Mx2jlkc365acQduuFwR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L151)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

It's currently set within `runTransferTimeoutHandler`. If we initialized in `processTransfer`, we're changing our definition of the timeout here - maybe that's OK. Previously, the transfer timeout starts right when we send the SHOW TRANSFER STATE query, which aligns well with the `stateTransferInProgress` state. With this change, it starts from when we called RequestTransfer successfully.
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 156 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-156:-Mx2jkd60XbjngHaQfmK:bjm5xxg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L156)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

That should work. How does the following sound: `stateTransferSessionSerialization`, and `stateTransferSessionDeserialization`? The latter includes trying to connect to a new SQL pod as well.
</blockquote></details>

Done. I also removed the connRecoverable sentinel entirely. We can be sure that once we're in the `stateTransferSessionDeserialization` state, the connection must be recoverable. Similarly, if we're in the `stateTransferSessionSerialization` state, the connection is non-recoverable.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 600 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r8-600:-Mx2LAfsEYW0So0U-p57:bf0qxg2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/forwarder.go#L600)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

We could definitely do that to make the API more robust, but I don't see a strong use case to that yet. The current intention is to reuse the existing selection algorithm for the connection migration work: https://github.com/cockroachdb/cockroach/blob/94c73132dba52afd20544c2792fdbdcd9ccddb77/pkg/ccl/sqlproxyccl/tenant/pod.go#L11-L23

That said, adding a _target_ pod makes it more complex compared to a _preferred_ pod since we now need to handle the case where the pod couldn't take in more connections. 
</blockquote></details>

Added this to the TODO comment.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 139 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwyybLiqoDcTTUX-fy3-r8-139:-Mx1n7tx7TmAamb4LX7L:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/metrics.go#L139)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I can do that. No strong preferences here.
</blockquote></details>

Done.
___
*[pkg/ccl/sqlproxyccl/metrics.go, line 151 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwyybLiqoDcTTUX-fy3-r8-151:-Mx2jwcI4DR-LabkzQSm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6e8dd554132e0eb46cb27fe50344dc16cfbb2f/pkg/ccl/sqlproxyccl/metrics.go#L151)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

Hm, let me look into this again. We may have to change where we increment the metric for this to work.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 158 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxD10gqAiZjwmwJhxMN:-MxD10gqAiZjwmwJhxMO:b-8cho8f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L158)):*
> ```Go
> 	// ------------------------------------------------------------------------
> 
> 	// transferKey is a unique string used to identify the transfer request,
> ```

@jeffswenson, how strongly do you feel that we need to guard these variables with a mutex when they don't need them? I've attempted to do so, but it was extremely difficult to read/follow, so I aborted the idea.  



_Code quote:_
```Go
	// transferKey is a unique string used to identify the transfer request,
	// and will be passed into the SHOW TRANSFER STATE statement. This will
	// be set to a randomly generated UUID whenever the transfer is requested
	// through the RequestTransfer API, and back to an empty string whenever the
	// transfer completes successfully or with a recoverable error.
	transferKey string

	// transferCtx has to be derived from ctx, and is created by the timeout
	// handler. All transfer related operations will use this so that they can
	// react to the timeout handler when that gets triggered.
	transferCtx context.Context

	// transferCloserCh is a channel that gets closed whenever the forwarder
	// transitions back to the ready state, which signifies that the transfer
	// process has completed successfully. Closing this will unblock the
	// client-to-server processor and stop the timeout handler.
	transferCloserCh chan struct{}
```


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
Have you made sure to run all available tests/scenarios using the Go `-race` detector? I always do this for any non-trivial multi-threaded code.

> Reviewed 17 of 17 files at r14, 9 of 9 files at r15.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MxD7wDs5Grd8TM44yhR:b-zhx21w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 224 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDB9RS7SggQFCrPbGl:-MxDB9RS7SggQFCrPbGm:bebyix8) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L224)):*
> ```Go
> 	// mainly used to unblock the client-to-server processor whenever the
> 	// forwarder has stopped, or a transfer has been requested.
> 	clientConn = pgwire.NewReadTimeoutConn(clientConn, func() error {
> ```

I'm a bit worried that we're "waking up" every second on every connection, since I expect the majority of connections to be otherwise idle. I'd expect that to have real overhead. Maybe if perf tests did show a lot of overhead, we would want to consider a 3rd goroutine to do the transfer...
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 372 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDAk6z5qS2ABoXzcU3:-MxDAk6z5qS2ABoXzcU4:b-1o8kri) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L372)):*
> ```Go
> 
> 				// When ForwardMsg gets blocked on Read, we will unblock that
> 				// through our custom readTimeoutConn wrapper.
> ```

I assume this custom wrapper will also allow unblocking the call to `PeekMsg` above, right? That seems to be the much more important case, since that might be blocked for very long periods of time simply b/c there are no client messages to read.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 410 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDQSux6ARjyt4TipTf:-MxDQSux6ARjyt4TipTg:b5uso2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L410)):*
> ```Go
> 			// finishTransfer will be called, so we should backup a local copy
> 			// first to prevent a race.
> 			key, closer := f.transferKey, f.transferCloserCh
> ```

While technically you're protected from race conditions by the `state` mutex, it's pretty non-standard to be reading fields on one goroutine without a lock when another goroutine will later update them. Typically you'd just put all fields that are read or written on different goroutines under the protection of a. mutex.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 663 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDSxWR37U2SNURSyYW:-MxDSxWR37U2SNURSyYX:b55phgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L663)):*
> ```Go
> 	// Updating the state also means that failures after this point are
> 	// recoverable (i.e. connections should not be terminated).
> 	f.mu.Lock()
> ```

NIT: You do this pattern in enough places that you should consider a `setState` helper method. That also lets you use the `defer` pattern, which is a best practice.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTRs!

OK. I have been mulling over this for a couple of days, and I like the new approach that I have. See last commit in https://github.com/cockroachdb/cockroach/pull/77419. That has a couple of major architectural change:
1. We no longer keep track of forwarder states.
2. Request and response processors have been split into their own internal components. We can suspend/resume these processors at any point in time by setting their appropriate deadlines. I took this idea from net/http: https://github.com/golang/go/blob/b5e06545b3c3b3b9c2ed8204e5ba3fd25a85db0f/src/net/http/server.go#L671-L723.
3. Due to the suspend/resume ability in (2), readTimeoutConn has been removed (This was pointed out in a previous comment about being slow. I didn't like this approach either, but couldn't seem to figure out a way to unblock the processors easily back then.)
4. All transfer related operations are now scoped in a single `runTransfer` method within `conn_migration.go`.
5. One major drawback to this design is that we incur more mutex calls during forwarding.

Thoughts?

I ran major tests, and they seem to work as intended:

```
=== RUN   TestConnectionMigration
    test_log_scope.go:79: test logs captured to: /var/folders/79/72cqh3ps56v7ts4h7zv92vz80000gq/T/logTestConnectionMigration4246808810
    test_log_scope.go:80: use -show-logs to present logs inline
=== RUN   TestConnectionMigration/successful
=== RUN   TestConnectionMigration/successful/normal_transfer
=== RUN   TestConnectionMigration/successful/failed_transfers_with_tx
=== RUN   TestConnectionMigration/successful/failed_transfers_with_dial_issues
=== RUN   TestConnectionMigration/transfer_timeout_in_response
=== CONT  TestConnectionMigration
    proxy_handler_test.go:1135: -- test log scope end --
--- PASS: TestConnectionMigration (12.97s)
    --- PASS: TestConnectionMigration/successful (8.92s)
        --- PASS: TestConnectionMigration/successful/normal_transfer (5.28s)
        --- PASS: TestConnectionMigration/successful/failed_transfers_with_tx (0.54s)
        --- PASS: TestConnectionMigration/successful/failed_transfers_with_dial_issues (3.03s)
    --- PASS: TestConnectionMigration/transfer_timeout_in_response (3.34s)
PASS
```

> Have you made sure to run all available tests/scenarios using the Go -race detector? I always do this for any non-trivial multi-threaded code.

Not explicitly. I relied on the stress/race test infra that we have on TeamCity. That did catch a couple of bugs, but they have all been addressed.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MxX0pYY9jRJX8K6ZO5K:bmhnmee)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 224 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDB9RS7SggQFCrPbGl:-MxX34ZQ-WFPHQEK0ezq:b-xk4kny) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L224)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm a bit worried that we're "waking up" every second on every connection, since I expect the majority of connections to be otherwise idle. I'd expect that to have real overhead. Maybe if perf tests did show a lot of overhead, we would want to consider a 3rd goroutine to do the transfer...
</blockquote></details>

Even if there was a 3rd goroutine, it would be tricky to do the transfer without unblocking the request processor. That said, I found a new approach to suspend/resume processors, so I think we should do that instead. One benefit to the 3rd goroutine is that we can now decouple most of the transferring logic from the forwarding logic.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 372 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDAk6z5qS2ABoXzcU3:-MxX3xNHCLVvU3ppF_3O:bbwp6kn) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L372)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I assume this custom wrapper will also allow unblocking the call to `PeekMsg` above, right? That seems to be the much more important case, since that might be blocked for very long periods of time simply b/c there are no client messages to read.
</blockquote></details>

Yes, that was the original intention of the readTimeoutConn wrapper, but as you've noted, we'd need to wake up every second for idle connections: https://github.com/cockroachdb/cockroach/blob/3b7f761f019c0457c19726bab8d14c231cd318fc/pkg/sql/pgwire/conn.go#L1800-L1804

Note that we're also using this same approach within the SQL server.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 410 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxDQSux6ARjyt4TipTf:-MxX4UWm03WLEmxSHjRr:b-gnvtha) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L410)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

While technically you're protected from race conditions by the `state` mutex, it's pretty non-standard to be reading fields on one goroutine without a lock when another goroutine will later update them. Typically you'd just put all fields that are read or written on different goroutines under the protection of a. mutex.
</blockquote></details>

Hm yeah, but the original intention was to avoid grabbing a mutex when we don't need it. I used a similar idea in the new approach on the `serverConn` field: https://github.com/cockroachdb/cockroach/pull/77419/commits/226d7d89bc69f9478391ee6e617cc3f69ecbdf3f#diff-f50d75f5a5e7b6d840c918b7ac35bb1f7605fca1f5434189c6eb99cd4bf67db2R397-R399. That said, if we wanted to use a mutex, that's fine too.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
I like the alternate approach, since idle connections incur no overhead - that seems to be the common case to optimize for. One thing I wasn't able to find when I took a look was the additional mutex calls during forwarding that you mention.





<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
>  One thing I wasn't able to find when I took a look was the additional mutex calls during forwarding that you mention.

That's within the request and response processors. Here's one example: https://github.com/cockroachdb/cockroach/blob/226d7d89bc69f9478391ee6e617cc3f69ecbdf3f/pkg/ccl/sqlproxyccl/forwarder.go#L260-L329

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-Mx_Yiro4gaonaTRGipb:b-pjc2qd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @JeffSwenson)



<!-- Sent from Reviewable.io -->

jaylim-crl committed:
ccl/sqlproxyccl: add last message details to the forwarder's processor

This commit adds last message details to the forwarder's processor, and in
particular, lastMessageType and lastMessageTransferredAt. The latter is
implemented using a simple logical clock, which will be used to determine
ordering of events. These two will be used during connection migration to
determine a safe transfer point.

At the same time, we plumb the connector and metrics into the processor, which
will be used in subsequent commits.

Release justification: Low risk, sqlproxy-only change.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: complete connection migration support in the forwarder

Informs cockroachdb#76000.

This commit completes the connection migration feature in the the forwarder
within sqlproxy. The idea is as described in the RFC.

A couple of new sqlproxy metrics have been added as well:
- proxy.conn_migration.success
- proxy.conn_migration.error_fatal
- proxy.conn_migration.error_recoverable
- proxy.conn_migration.attempted

For more details, see metrics.go in the sqlproxyccl package.

Release justification: This completes the first half of the connection
migration feature. This is low risk as part of the code is guarded behind the
connection migration feature, which is currently not being used in production.
To add on, CockroachCloud is the only user of sqlproxy.

Release note: None

jaylim-crl committed:
ccl/sqlproxyccl: support waitResumed on the processor to block until resumption

Previously, there could be a race where suspend() was called right after
resuming the processors. If the processor goroutines have not started, suspend
will implicitly return, leading to a violation of an invariant, where we want
the processors to be suspended before proceeding. This commit adds a new
waitResumed method on the processor that allows callers to block until the
processors have been resumed.

Release justification: sqlproxy-only change.

Release note: None

jaylim-crl reviewed (commented):
OK - this is ready for another round of review. Some tests are TODO, which will be addressed before merging, but I'm putting this up for a review first. Only look at the last three commits. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MxpJDvk7wzE4j9vnawy:b-vqalnn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @JeffSwenson)



<!-- Sent from Reviewable.io -->

JeffSwenson commented on pkg/ccl/sqlproxyccl/conn_migration.go:
```diff
@@ -13,13 +13,277 @@ import (
+		}
+	}()
+
+	f.mu.Lock()
```
This mutex is held for most of the transfer process. The processors are thread safe, and there is the isTransferring state, so we can avoid most of this locking if we create helpers for retrieving and setting the processors.

The code as written is correct. I like to keep critical sections relatively short in an effort to reduce the probability of deadlocks and contention.

I'm thinking something like:
```
func (f* forwarder) getProcessors() (request processor, response processor)  {
 	f.mu.Lock()
	defer f.mu.Unlock()
	return f.mu.request, f.mu.response
}

func (f* forwarder) replaceServerConn(newServerConn net.Conn) {
	f.mu.Lock()
	defer f.mu.Unlock()
	// Transfer was successful.
	clockFn := makeLogicalClockFn()
	f.mu.serverConn.Close()
	f.mu.serverConn = newServerConn
	f.mu.request = newProcessor(clockFn, f.mu.clientConn, f.mu.serverConn)
	f.mu.response = newProcessor(clockFn, f.mu.serverConn, f.mu.clientConn)
}

...

request, response := getProcessors()    

// Suspend both processors before starting the transfer.
if err := request.suspend(ctx); err != nil {
    return errors.Wrap(err, "suspending request processor")
}
if err := response.suspend(ctx); err != nil {
    return errors.Wrap(err, "suspending response processor")
}

// Transfer the connection.
newServerConn, err := transferConnection(ctx, f.connector, f.mu.clientConn, f.mu.serverConn)
if err != nil {
    return errors.Wrap(err, "transferring connection")
}
f.replaceServerConn(newServerConn)
	
return nil
```

jeffswenson reviewed (commented):
This is looking great! I have a few minor comments.

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MxtlHUIEziab0--STko:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/conn_migration.go, line 125 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwhaaVoixzBmJ99wZRJ-r22-125:-Mxtj8e9-nX-pW435k8E:b7lmjx3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9661512a35f971b7335c929ea9d6c5f7e6ae5ad8/pkg/ccl/sqlproxyccl/conn_migration.go#L125)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

I don't think logCtx is necessary. I looked at the implementation of log and I did not find any calls that check ctx.Err(). If it did, I would argue that is a bug. It is normal to log something that goes wrong during the cancellation process.
</blockquote></details>

I didn't have this initially, but the process panics with a span-use-after-close error:

```
panic: use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 2224 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xb7faeb8)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:185 +0x134
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc0011d2900)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:387 +0x1e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:395
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x8f543d0, 0xc005301580})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x3a
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth({0x8f543d0, 0xc005301580}, 0x1, 0x1, 0x0, {0x7d9f468, 0x1f}, {0xc0051a7d60, 0x1, 0x1})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/channels.go:55 +0x2a8
github.com/cockroachdb/cockroach/pkg/util/log.Infof(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/log_channels_generated.go:263
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).runTransfer.func2()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:145 +0x1f8
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).runTransfer(0xc00517d200)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:174 +0x5d7
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).RequestTransfer.func1()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/forwarder.go:154 +0x1d
created by github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).RequestTransfer
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/forwarder.go:154 +0x5b
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl	2.852s
```
___
*[pkg/ccl/sqlproxyccl/conn_migration.go, line 148 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwhaaVoixzBmJ99wZRJ-r22-148:-MxtlRAT3Tu4V83BmYHj:brt6y6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/9661512a35f971b7335c929ea9d6c5f7e6ae5ad8/pkg/ccl/sqlproxyccl/conn_migration.go#L148)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

This mutex is held for most of the transfer process. The processors are thread safe, and there is the isTransferring state, so we can avoid most of this locking if we create helpers for retrieving and setting the processors.

The code as written is correct. I like to keep critical sections relatively short in an effort to reduce the probability of deadlocks and contention.

I'm thinking something like:
```
func (f* forwarder) getProcessors() (request processor, response processor)  {
 	f.mu.Lock()
	defer f.mu.Unlock()
	return f.mu.request, f.mu.response
}

func (f* forwarder) replaceServerConn(newServerConn net.Conn) {
	f.mu.Lock()
	defer f.mu.Unlock()
	// Transfer was successful.
	clockFn := makeLogicalClockFn()
	f.mu.serverConn.Close()
	f.mu.serverConn = newServerConn
	f.mu.request = newProcessor(clockFn, f.mu.clientConn, f.mu.serverConn)
	f.mu.response = newProcessor(clockFn, f.mu.serverConn, f.mu.clientConn)
}

...

request, response := getProcessors()    

// Suspend both processors before starting the transfer.
if err := request.suspend(ctx); err != nil {
    return errors.Wrap(err, "suspending request processor")
}
if err := response.suspend(ctx); err != nil {
    return errors.Wrap(err, "suspending response processor")
}

// Transfer the connection.
newServerConn, err := transferConnection(ctx, f.connector, f.mu.clientConn, f.mu.serverConn)
if err != nil {
    return errors.Wrap(err, "transferring connection")
}
f.replaceServerConn(newServerConn)
	
return nil
```


</blockquote></details>

Done. Updated it to your suggestion, but also introduced `getConns` for the client and server connections.
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 190 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r18-190:-MxtkJGx8AMQQXnaM5o-:bjlxkcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f52149ca576ebf7bff35837961c8a2e7d025bb3/pkg/ccl/sqlproxyccl/forwarder.go#L190)):*
<details><summary><i>Previously, JeffSwenson (Jeff Swenson) wrote…</i></summary><blockquote>

:smile: On a 100gbps ethernet line, sending the smallest possible pg message, with no protocol overhead, it would take 217 years of continuous transmission to overflow an int64 counter.

```
>>> line_rate = 100 * 2 ** 30 # 100 gibs
>>> message_size = 5 * 8 # minimum message size of 8 bytes
>>> msg_per_second = line_rate / message_size
>>> seconds_to_overflow = 2 ** 64 / msg_per_second 
>>> years_to_overflow = seconds_to_overflow / (60 * 60 * 24 * 365)
>>> years_to_overflow
217.908031253171
```
</blockquote></details>

:smile:
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 320 at r18](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r18-320:-Mxtkofn1jP5-KTajM3I:b-5kwmzn) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f52149ca576ebf7bff35837961c8a2e7d025bb3/pkg/ccl/sqlproxyccl/forwarder.go#L320)):*
> It would require passing peekErr and typ as arguments.

Doing this muddies the logic in exitPeek, and I'd rather keep it this way with the extra mutex lock+unlock. If this proves to be an issue, we could easily optimize this at a later time. Doing it the current way also makes sure that updating the last message details is tied to the ForwardMsg call, and not when exiting. 
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 83 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwFOQDfl9Veg40qOiuL-r20-83:-MxtjkHI0TvboXC12KSl:bylj3r9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fddfd308ec3106e1be705dbd26c99e38fb6efa26/pkg/ccl/sqlproxyccl/forwarder.go#L83)):*
Done. Added a note about that.

> Generally I would advise against acquiring multiple locks at the same time because it is a deadlock risk. 

I think as long as we're aware about it, such risks can be minimized. Also, using a single lock for all three structs is going to cause contention between the request and response processors especially when queries are pipelined.


<!-- Sent from Reviewable.io -->

jeffswenson reviewed (approved):
LGTM!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-MxuMYAkCOeX7W9sbGJw:b-85i648)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @jaylim-crl, and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/conn_migration.go, line 125 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwhaaVoixzBmJ99wZRJ-r22-125:-MxuENJzCHc-N7Vft5Fg:bfafyk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/9661512a35f971b7335c929ea9d6c5f7e6ae5ad8/pkg/ccl/sqlproxyccl/conn_migration.go#L125)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

I didn't have this initially, but the process panics with a span-use-after-close error:

```
panic: use of Span after Finish. Span: proxy-con-serve. Finish previously called at: <stack not captured. Set debugUseAfterFinish>

goroutine 2224 [running]:
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).detectUseAfterFinish(0xb7faeb8)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:185 +0x134
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).RecordingType(0xc0011d2900)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:387 +0x1e
github.com/cockroachdb/cockroach/pkg/util/tracing.(*Span).IsVerbose(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/tracing/span.go:395
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog({0x8f543d0, 0xc005301580})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:92 +0x3a
github.com/cockroachdb/cockroach/pkg/util/log.logfDepth({0x8f543d0, 0xc005301580}, 0x1, 0x1, 0x0, {0x7d9f468, 0x1f}, {0xc0051a7d60, 0x1, 0x1})
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/channels.go:55 +0x2a8
github.com/cockroachdb/cockroach/pkg/util/log.Infof(...)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/util/log/log_channels_generated.go:263
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).runTransfer.func2()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:145 +0x1f8
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).runTransfer(0xc00517d200)
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/conn_migration.go:174 +0x5d7
github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).RequestTransfer.func1()
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/forwarder.go:154 +0x1d
created by github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl.(*forwarder).RequestTransfer
	/Users/jaylim/workspace/src/github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl/forwarder.go:154 +0x5b
FAIL	github.com/cockroachdb/cockroach/pkg/ccl/sqlproxyccl	2.852s
```
</blockquote></details>

Jay and I dug into this off line. The root cause is the sqlproxy is using stopper incorrectly. This will be fixed in a separate PR. https://github.com/cockroachdb/cockroach/issues/77689
___
*[pkg/ccl/sqlproxyccl/forwarder.go, line 158 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MxD10gqAiZjwmwJhxMN:-MxuMI-k8svjmvfj2WXy:b-mp783r) ([raw file](https://github.com/cockroachdb/cockroach/blob/2fdda2dd701992fd2cd632542472cd619e04e922/pkg/ccl/sqlproxyccl/forwarder.go#L158)):*
<details><summary><i>Previously, jaylim-crl (Jay Lim) wrote…</i></summary><blockquote>

@jeffswenson, how strongly do you feel that we need to guard these variables with a mutex when they don't need them? I've attempted to do so, but it was extremely difficult to read/follow, so I aborted the idea.  
</blockquote></details>

This was resolved by the redesign.


<!-- Sent from Reviewable.io -->

jaylim-crl reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-:-My0_1hp1x8g71ZS1iPW:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball and @JeffSwenson)

___
*[pkg/ccl/sqlproxyccl/conn_migration.go, line 125 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/76805#-MwhaaVoixzBmJ99wZRJ-r22-125:-My0_3kI-cpAygRx-zH6:bkqdy3x) ([raw file](https://github.com/cockroachdb/cockroach/blob/9661512a35f971b7335c929ea9d6c5f7e6ae5ad8/pkg/ccl/sqlproxyccl/conn_migration.go#L125)):*
> The root cause is the sqlproxy is using stopper incorrectly. 

I guess it's more like we're not using the stopper entirely. Some places use `stopper.RunAsyncTask`, whereas others use the `go` keyword directly. The latter is handy as we don't need to thread the stopper all the way down. Fixing all uses of the `go` keyword isn't straightforward. We'd also have to rethink how we want traces to be displayed.


<!-- Sent from Reviewable.io -->

jaylim-crl commented:
bors r=JeffSwenson
